### ISSUE TITLE
Fuzzy partial fingerprint matching and fingerprint versioning

### ISSUE THREAD

**Issue by Evmaus-MS** [oasis-tcs/sarif-spec#164]
In the SARIF Spec, we cover some of the usage of fingerprints by result management systems.  We also have a namespacing concept.

There are two scenarios I think we may want to provide guidance on or a convention for when providing fingerprints for a tool:

First, we should provide some guidance on versioning fingerprints, and how a result management system should read a versioned fingerprint--e.x. if you have an older fingerprinting algorithm and make some improvements to it in a tool update, you will want to be able to say "If 'Fingerprint Version 2' is present, prefer matching on it.  Otherwise, match on 'Fingerprint Version 1'".  My suggestion would be to use a particular namespacing convention, e.x. the tool should output "FingerprintName/v1", "FingerprintName/v2" if the results management system should interpret the fingerprints in that way.  We may also want to suggest that a results management system should treat "FingerprintName" as "FingerprintName/v1" so that version information can be added after the original tool iteration.

Second, it seems possible that a tool may want to provide multiple values, of which "at least one" should match.  An example of this would be providing context before and after a result, where only one of the two is necessary for the partial fingerprint to be considered a match.  (Consider Semmle's hash based fingerprinting algorithm described in section 3.C here:  https://semmle.com/wp-content/uploads/2015/01/team-insight.pdf ).  I'd suggest we again use namespacing conventions for this, perhaps allowing "FingerprintName/0", "FingerprintName/1", "FingerprintName/2"--where any one of the three can match for "FingerprintName" to be considered a match.

We'd also probably want to cover the appropriate way to combine them (do you version and then group, or group and then version, and what's the meaning of either).

This may fall into the "guidance on usage of SARIF" rather than the specification itself, but this seems like something it'd be good to at least give some advice on.

**Comment by michaelcfanning** [comment]
@lgolding I don't understand the net outcome here that you are proposing. Are you agreeing that we will provide versioning for partial fingerprints. I am also not sure myself that we need to do anything with fingerprints. The fingerprints data are final, computed unique (or aspirationally unique) identifiers. If you have a match for any of them, you presumably have matched the result.

**Comment by lgolding** [comment]
**TL;DR**:

I would agree that both partial fingerprint names and (full) fingerprint names could have a "version" component. I don't agree with the notion of "grouping" partial fingerprints.

**Details**:

It's possible that Everett's original issue description wasn't precise in its use of the terms "fingerprint" and "partial fingerprint". I was responding to what I took to be his proposal, based on the terminology he used:
- Allow a "version" component on fingerprint names.
- Allow a "grouping" component on partial fingerprint names.

With that in mind, here is my (hopefully clarified) response:

1. Fingerprints are indeed final, computed values that reliably distinguish logically distinct results. But there might have more than one fingerprinting algorithm, and each one might go through versions. So I was agreeing to what I _thought_ Everett was proposing, namely, to allow the property names in the `result.fingerprints` object to have a "version" component (_e.g_, `"myAlgorithm/v1"`, `"yourAlgorithm/v2"`).

2. At the same time, I was pushing back on the notion of naming _related_ partial fingerprints with a common base name. I _thought_ Everett was proposing to structure the property names in the `result.partialFingerprints` object like this: `"regionHash/1"`, `"regionHash/2"`. I argued that I didn't see why that was better than `"precedingRegionHash"`, `"followingRegionHash"`.

3. As to whether to _version_ partial fingerprints: I think it less likely that a partial fingerprint would need to be versioned than that a (full) fingerprint would need to be versioned. Our canonical example for a partial fingerprint -- something that a result management system couldn't infer from the contents of a log file -- is a "forbidden word" detected by a geopolitical/cultural sensitivity checker. The result might be
```
{
  "ruleId": "POL0001",
  "message": {
   "text": "Don't use the word 'foo'."
  },
  "partialFingerprints": {
    "forbiddenWord": "foo"
  }
}
```
But I suppose even partial fingerprint algorithms _could_ be versioned. For example, you might decide that your `"precedingRegionHash"` partial fingerprint should include the preceding 5 lines instead of the preceding 10 lines. So sure, I'd allow a version component in partial fingerprint names.

**Comment by lgolding** [comment]
@Evmaus-MS @michaelcfanning 

To my great surprise, the SARIF v2 spec _already_ supports your versioned fingerprints! See §3.19.10, "fingerprints property":

> Each property name in this object **SHALL** be a hierarchical string (§3.4.4). A result management system **MAY** use the property names to identify the method used to calculate the fingerprint.
>
> EXAMPLE: In this example, the producer has calculated a fingerprint using version 2 of a fingerprinting method it refers to as "contextRegionHash":
>```
> {
>    "fingerprints": {
>      "contextRegionHash/v2": "097886bc876fe"
>    }
>}
>```

**Comment by lgolding** [comment]
@Evmaus-MS @michaelcfanning

And §3.19.11 says the same thing for `result.partialFingerprints`.

**Comment by michaelcfanning** [comment]
Yay! we're done. btw, Larry, if you're interested, I can provide a clear example of the value of grouped partial fingerprints offline (based on some practical experimentation in our emerging SARIF pipeline).

**Comment by lgolding** [comment]
@michaelcfanning Let me clearly understand what you mean by "done". My response to Everett's proposal was:

1. We should allow versioned fingerprints --**DONE**
2. We should allow versioned partial fingerprints -- **DONE**
3. We should _not_ bother with "grouped" partial fingerprints.

So if you agree with all three of my points (in particular #&#xFEFF;3, we are indeed done.

But you just said that you _do_ see value in "grouped" partial fingerprints. If so, we are _not_ done, right?

**Comment by lgolding** [comment]
@michaelcfanning ... well, I suppose that once you allow hierarchical names in `result.fingerprints` and `result.fingerprints`, you can use them for anything, including grouping, for example:
```
regionHash/preceding/v1
regionHash/following/v2
```
So in that sense, we are indeed done. But @Evmaus-MS was asking for guidance on the naming convention, and on how to combine the components of the hierarchical name:
> We'd also probably want to cover the appropriate way to combine them (do you version and then group, or group and then version, and what's the meaning of either).
So in that sense, we're not done.

**Comment by lgolding** [comment]
@michaelcfanning @Evmaus-MS 

My proposal:
- Property names in both fingerprints and partial fingerprints are hierarchical (spec already says that).
- If the _last_ component is in the form `vn` where n is a positive integer, it is interpreted as a version number.

**Comment by michaelcfanning** [comment]
Yes

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
